src/compiler/moduleSpecifiers.ts

```diff
@@ -832,17 +847,54 @@ const enum MatchingMode {
     Pattern,
 }
 
-function tryGetModuleNameFromExports(options: CompilerOptions, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode = MatchingMode.Exact): { moduleFileToTry: string; } | undefined {
+function getOutputPath(targetFilePath: string, options: CompilerOptions, commonDir: string) {
+    if (options.emitDeclarationOnly) {
+        return;
+    }
+    return changeExtension(
+        options.outDir ?
+            resolvePath(
+                options.outDir,
+                getRelativePathFromDirectory(commonDir, targetFilePath, /*ignoreCase*/ false),
+            ) :
+            targetFilePath,
+        getOutputExtension(targetFilePath, options),
+    );
+}
+
+function tryGetModuleNameFromExportsOrImports(options: CompilerOptions, host: ModuleSpecifierResolutionHost, getCanonicalFileName: GetCanonicalFileName, targetFilePath: string, packageDirectory: string, packageName: string, exports: unknown, conditions: string[], mode: MatchingMode, isImports: boolean): { moduleFileToTry: string; } | undefined {
     if (typeof exports === "string") {
+        const currentDirectory = host.getCurrentDirectory();
+        const commonDir = getNormalizedAbsolutePath(
+            getCommonSourceDirectory(
+                options,
+                () => [],
+                currentDirectory,
+                getCanonicalFileName,
+            ),
+            currentDirectory,
+        );
+
+        const outputFile = getOutputPath(targetFilePath, options, commonDir);
+
         const pathOrPattern = getNormalizedAbsolutePath(combinePaths(packageDirectory, exports), /*currentDirectory*/ undefined);
         const extensionSwappedTarget = hasTSFileExtension(targetFilePath) ? removeFileExtension(targetFilePath) + tryGetJSExtensionForFile(targetFilePath, options) : undefined;
+
         switch (mode) {
             case MatchingMode.Exact:
-                if (comparePaths(targetFilePath, pathOrPattern) === Comparison.EqualTo || (extensionSwappedTarget && comparePaths(extensionSwappedTarget, pathOrPattern) === Comparison.EqualTo)) {
+                if (outputFile && comparePaths(outputFile, pathOrPattern) === Comparison.EqualTo || comparePaths(targetFilePath, pathOrPattern) === Comparison.EqualTo || (extensionSwappedTarget && comparePaths(extensionSwappedTarget, pathOrPattern) === Comparison.EqualTo)) {
                     return { moduleFileToTry: packageName };
                 }
                 break;
             case MatchingMode.Directory:
+                if (outputFile && containsPath(pathOrPattern, outputFile)) {
+                    const fragment = getRelativePathFromDirectory(pathOrPattern, outputFile, /*ignoreCase*/ false);
+                    return { moduleFileToTry: combinePaths(packageName, fragment) };

```

Comment:
This is using a different pattern~ for the returned value than other branches in this case. I think that it's correct and that actually those other ones might be broken but I didn't yet have time to recheck that.